src/main/java/in/zapr/druid/druidry/client/DruidJerseyClient.java

```diff
@@ -0,0 +1,125 @@
+
+        log.debug("Will query to {}", druidUrl);
+        this.druidConfiguration = druidConfiguration;
+    }
+
+    @Override
+    public void connect() throws ConnectionException {
+
+        PoolingHttpClientConnectionManager connectionManager
+                = new PoolingHttpClientConnectionManager();
+
+        int numberOfConnectionsInPool = DEFAULT_CONNECTION_POOL_LIMIT;
+
+        if (this.druidConfiguration.getConcurrentConnectionsRequired() != null) {
+            numberOfConnectionsInPool = this.druidConfiguration.getConcurrentConnectionsRequired();
+        }
+
+        connectionManager.setDefaultMaxPerRoute(numberOfConnectionsInPool);
+
+        try {
+            if (jerseyConfig == null) {
+                this.jerseyConfig = new ClientConfig();
+            }
+
+            this.jerseyConfig.property(ApacheClientProperties.CONNECTION_MANAGER, connectionManager);
+            this.jerseyConfig.connectorProvider(new ApacheConnectorProvider());
+
+            this.client = ClientBuilder.newClient(this.jerseyConfig);
+            this.queryWebTarget = this.client.target(this.druidUrl);
+
+        } catch (Exception e) {
+            throw new ConnectionException(e);
+        }
+    }
+
+    @Override
+    public void close() throws ConnectionException {
+        try {
+            this.client.close();
+        } catch (Exception e) {
+            throw new ConnectionException(e);
+        }
+    }
+
+    @Override
+    public String query(DruidQuery druidQuery) throws QueryException {
+
+        try (Response response = this.queryWebTarget
+                .request(MediaType.APPLICATION_JSON)
+                .post(Entity.entity(druidQuery, MediaType.APPLICATION_JSON))) {
+
+            if (response.getStatus() == Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()) {
+                handleInternalServerResponse(response);
+            }
+
+            return response.readEntity(String.class);
+
+        } catch (Exception e) {
+            log.error("Exception while querying {}", e);
+            throw new QueryException(e);
+        }
+    }
+
+    @Override
+    public <T> List<T> query(DruidQuery druidQuery, Class<T> className) throws QueryException {
+        try (Response response = this.queryWebTarget
+                .request(MediaType.APPLICATION_JSON)
+                .post(Entity.entity(druidQuery, MediaType.APPLICATION_JSON))) {
+
+            if (response.getStatus() == Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()) {
+                handleInternalServerResponse(response);
+            }
+
+            return response.readEntity(new GenericType<List<T>>() {
+            });
+        } catch (Exception e) {
+            log.error("Exception while querying {}", e);
+            throw new QueryException(e);
+        }
+    }
+
+    private void handleInternalServerResponse(Response response) throws QueryException {
+        String message = response.readEntity(String.class);
```

Comment:
In case of 500 error, Druid always returns a string message? Is that not a standard json? If it is a standard json in any error case, we could serialize it to corresponding model class

Reply:
It returns a standard JSON, but lets discuss this once

Reply:
If it a standard json, can we discuss on this? I have a few suggestions.